datadog_checks_dev/datadog_checks/dev/tooling/manifest_validator/common/validator.py

```diff
@@ -177,36 +178,86 @@ def validate(self, check_name, decoded, _):
+        ),
+    }
+
+    SHORT_NAME_PATHS = {
+        V1: (
+            "assets/dashboards",
+            "assets/monitors",
+            "assets/saved_views",
+        ),
+        V2: (
+            "assets/dashboards",
+            "assets/monitors",
+            "assets/saved_views",
+        ),
     }
 
     def validate(self, check_name, decoded, fix):
-        manifest_fields_changed = content_changed(file_glob=f"{check_name}/manifest.json")
+        # Check if previous version of manifest exists
+        # If not, this is a new file so this validation is skipped
+        try:
+            previous = git_show_file(path=f"{check_name}/manifest.json", ref="master")
+            previous_manifest = JSONDict(json.loads(previous))
+        except Exception:
+            self.result.messages['info'].append(
+                "  skipping check for changed fields: integration not on default branch"
+            )
+            return
 
         # Skip this validation if the manifest was updated from 1.0.0 -> 2.0.0
-        manifest_upgrade_regex = re.compile(r'\+\s+"manifest_version": "2.0.0",')
-        version_upgraded = manifest_upgrade_regex.findall(manifest_fields_changed)
+        current_manifest = decoded
+        if (
+            previous_manifest[self.MANIFEST_VERSION_PATH] == "1.0.0"
+            and current_manifest[self.MANIFEST_VERSION_PATH] == "2.0.0"
+        ):
+            self.result.messages['info'].append("  skipping check for changed fields: manifest version was upgraded")
+            return
 
-        if 'new file' not in manifest_fields_changed and not version_upgraded:
-            for field in self.FIELDS_NOT_ALLOWED_TO_CHANGE[self.version]:
-                if field in manifest_fields_changed:
-                    output = f'Attribute `{field}` is not allowed to be modified. Please revert to original value'
+        # Check for differences in immutable attributes
+        immutable_attribute_changed = False
+        for key_path in self.IMMUTABLE_FIELD_PATHS[self.version]:
+            previous_value = previous_manifest.get_path(key_path)
+            current_value = current_manifest.get_path(key_path)
+
+            if previous_value != current_value:
+                output = f'Attribute `{current_value}` at `{key_path}` is not allowed to be modified. Please revert it \
+to the original value `{previous_value}`.'
+                immutable_attribute_changed = True
+                self.fail(output)
+
+        # If already failed, stop validations with failure messages gathered above
+        if immutable_attribute_changed:
+            return
+
+        # Check for differences in `short_name` keys
+        for key_path in self.SHORT_NAME_PATHS[self.version]:
+            previous_short_name_dict = previous_manifest.get_path(key_path) or {}
+            current_short_name_dict = current_manifest.get_path(key_path) or {}
+
+            # Every `short_name` in the prior manifest must be in the current manifest
```

Comment:
This implies we cannot remove or replace these assets from a manifest, which I'm not sure we want.  Perhaps we need to ensure that previous (name, asset) pairs remain consistent, instead.

Reply:
Retracting my comment - forgot we can't delete these assets once they have been synced.